src/core/server/server.ts

```diff
@@ -61,16 +64,19 @@ export class Server {
     const httpSetup = await this.http.setup();
     this.registerDefaultRoute(httpSetup);
 
+    const contextServiceSetup = this.context.setup();
     const elasticsearchServiceSetup = await this.elasticsearch.setup({
       http: httpSetup,
     });
 
     const pluginsSetup = await this.plugins.setup({
+      context: contextServiceSetup,
```

Comment:
This is true and I'm not sure how to change this without introducing yet another lifecycle method so that plugins could be discovered before anything else sets up. It may be worth just making an exception for the plugin discovery and making that run before any lifecycle methods without introducing a new method to the entire platform.

Separately, you could argue that the context pattern exists mostly to support using functionality available during `start` inside a function registered during `setup`. In that case, we probably won't _need_ to register contexts during `setup` but it is a strange limitation.

I'm supportive of moving plugin discovery so that it runs before the rest of the systems get set up.

Reply:
I've implemented this option in a new commit in this PR: https://github.com/elastic/kibana/pull/42395/commits/00f67c4cccf7e440b4b2dea30b3c099268c30bbd